### ISSUE TITLE
Add ProcessorParam Preprocessor Support

### ISSUE THREAD

**Issue by dellis1972** [MonoGame/MonoGame#4417]
This might be extremely handy when targeting multiple GPU chip sets on android. 
Because android supports a ton of compression formats the chances are multiple .apk's would be required. Rather than having multiple .mgcb files it would be nice to be able to do

/processorParam:TextureFormat=$(TextureFormat)

and then provide that either via the command line

/define:Textureformat=Compressed

or (more likely) via the following in the .mcgb file itself and pass the /config parameter.

$if config=PVRTC
$set TextureFormat=PVRCompressed
$endif

Then replace the $(TextureFormat) of the ProcessorParam when its parsed.
At the moment the following is required for each which can be a bit of a pain when dealing with multiple compression types.

$if config=PVRTC
/processorParam:TextureFormat=PVRCompressed
$endif
$if config=ATI
/processorParam:TextureFormat=ATICompressed
$endif

Comments?

**Comment by jamesford42** [comment]
/define:Textureformat=Compressed
...
/processorParam:TextureFormat=$(TextureFormat)

This seems like a valuable and simple to do extension to the preprocessor. You might take a peek at Tools/MGCB/CommandLineParser.cs, that is where it would be implemented.

**Comment by dellis1972** [comment]
@jamesford42 yup :) I've been in there once or twice before, just wanted to log this incase someone can implement it before I do :D

**Comment by tomspilman** [comment]
Having some sort of preprocessor feature like this isn't a bad idea...  but not for the use suggested here.

I feel the change to specify specific formats via `TextureFormat` is a bad path to head down.  The whole point of the pipeline is for it to package content in appropriate target formats for the platform.  If the developer has to start choosing formats for things it will become a mess to manage your content when targeting 5 or 6 platforms like some games are doing.

We need to have a separate conversation... possibly one we've had before I've forgotten about...  on how to deal with the Android texture mess.  We really should not live with it as it is now.

**Comment by dellis1972** [comment]
@tomspilman by default the "Compressed" option on android will to ETC1 for Opaque images and 16bit for Alpha  that will work everywhere, that was the plan we cam up with :) BUT the compression won't be as good as it can be. 

Google allow devs to upload packages for each Texture compression type (among other things). So you can do a DXT build then a PVRTC build then a ATITC build. However I don't think its a fantastic idea to have to maintain 3/4 .mgcb files all of which have the same content except a few processor options. 

I worked around the issue for now using the /config option as mentioned above so I can then inject that value into the <Header> element of the Content targets file and pick the correct compression format, but its not idea. I'm just wondering if we can make this easier for developers somehow, building 3-4 .apks for google is a pain in itself without the overhead of maintaining a bunch of content files on top.

On a side note I think the Pipeline tool deletes any custom $if stuff if you make modifications in the editor to an existing .mgcb file that has been hand edited..we probably should fix that at some point :)

**Comment by tomspilman** [comment]
> Google allow devs to upload packages for each Texture compression type 

I think maybe we need to introduce the concept of multiple packages to content building... something XNA never considered.

I see three, maybe four cases where the developer might want multiple content packages:

  - Texture format mess (Android).
  - Content size reduction for lower end or smaller screen devices (Android, iOS, UWP).
  - Content changes based on target ie. phone vs tablet vs pc (Android, iOS, UWP).
  - Different content for trial modes, demos, etc. (all platforms).

Assuming we can generalize some of this into a few properties on the .mgcb project, when you build your content it would process some assets multiple times to target the difference cases defined at the project level.  You then end up with multiple content folders targeting these different situations based on your preferences and target platform.

This way, instead of you having to find only your alpha textures and set it to export some specific format for Android, the pipeline simply builds all three for you when appropriate and organizes them so that you can easily do the multi-package thing.  It can even do something smart like only importing the texture once, then copy and process it multiple times.  If you then build for Windows UWP, it knows it doesn't have this mess to deal with, and simply exports the one texture format needed.

To me this seems like a better general approach to the issue and makes it work like the pipeline was intended...  just tell it what platform to target and the right stuff is built.

**Comment by tomspilman** [comment]
> On a side note I think the Pipeline tool deletes any custom $if stuff if you make 

Yeah...  we basically need to use a real parser on the .mgcb file like we do for the MGFX parsing.  A real parser generates a bunch of tokens which we can process and replace.  This would let us update an existing file and resave it without loosing custom edits around it.

**Comment by dellis1972** [comment]
This was implemented in a different way via https://github.com/MonoGame/MonoGame/commit/adacb8689ab5d6607e5452da381e2fc0a1542394
